components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/listener/MyAccountApplicationCreationListener.java

```diff
@@ -0,0 +1,193 @@
+import org.wso2.carbon.identity.role.v2.mgt.core.exception.IdentityRoleManagementException;
+import org.wso2.carbon.identity.role.v2.mgt.core.model.Permission;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.wso2.carbon.identity.api.resource.mgt.constant.APIResourceManagementConstants.APIResourceTypes.TENANT;
+import static org.wso2.carbon.identity.api.resource.mgt.constant.APIResourceManagementConstants.RBAC_AUTHORIZATION;
+import static org.wso2.carbon.identity.application.mgt.ApplicationConstants.IMPERSONATE_ROLE_NAME;
+import static org.wso2.carbon.identity.application.mgt.ApplicationConstants.IMPERSONATE_SCOPE_NAME;
+import static org.wso2.carbon.identity.application.mgt.ApplicationConstants.MY_ACCOUNT_APPLICATION_NAME;
+import static org.wso2.carbon.identity.application.mgt.ApplicationConstants.USER_SESSION_IMPERSONATION_ENABLED;
+
+/**
+ * Listener to handle post tasks of My Account application creation.
+ */
+public class MyAccountApplicationCreationListener extends AbstractApplicationMgtListener {
+
+    private final AuthorizedAPIManagementService authorizedAPIManagementService = new
+            AuthorizedAPIManagementServiceImpl();
+
+    @Override
+    public int getExecutionOrderId() {
+
+        return 1;
+    }
+
+    @Override
+    public int getDefaultOrderId() {
+
+        return 1;
+    }
+
+    @Override
+    public boolean isEnable() {
+
+        return Boolean.parseBoolean(IdentityUtil.getProperty(USER_SESSION_IMPERSONATION_ENABLED));
+    }
+
+    @Override
+    public boolean doPostCreateApplication(ServiceProvider serviceProvider, String tenantDomain, String userName)
+            throws IdentityApplicationManagementException {
+
+        String appName = serviceProvider.getApplicationName();
+        if (!MY_ACCOUNT_APPLICATION_NAME.equals(appName)) {
+            return true;
+        }
+        String myAccountId = getMyAccountAppId(tenantDomain);
+        // Register API Resource.
+        if (!isAPIResourceRegistered(myAccountId, tenantDomain)) {
+            addApiResourceToApplication(myAccountId, tenantDomain);
+        }
+        // Create Role.
+        if (!isExistingRoleName(myAccountId, tenantDomain)) {
+            addImpersonatorRole(myAccountId, tenantDomain);
+        }
+
+        return true;
+    }
+
+    private void addImpersonatorRole(String myAccountId, String tenantDomain)
+            throws IdentityApplicationManagementException {
+
+        List<Permission> permissions = new ArrayList<>();
+        Permission permission = new Permission(IMPERSONATE_SCOPE_NAME);
+        permissions.add(permission);
+
+        RoleManagementService roleManagementService = ApplicationManagementServiceComponentHolder.getInstance()
+                .getRoleManagementServiceV2();
+        try {
+            roleManagementService.addRole(IMPERSONATE_ROLE_NAME, new ArrayList<>(), new ArrayList<>(), permissions,
+                    "application", myAccountId, tenantDomain);
```

Comment:
Shall we define "application" as a constant?

Reply:
Addressed with 82654fc5ad702dedb715d648bde39c8f8b15aa16